### PR TITLE
Closes #268 easier user experience for splitting datasets

### LINKED ISSUE #268: Feature Request: `xportr_split()` enhancement for easier user experience
### Feature Idea

@bms63 @elimillera  - i chatted with our esub team here further about the split datasets (and showed them the new `xportr_split()` function). they had 2 pieces of feedback to enhance this function:

- users currently either have to manually work out if this is needed or check the warning from `xportr_write()` which is not flexible enough (see https://github.com/atorus-research/xportr/issues/267 - as China limit is 4GB and Japan is 10GB for example). their advice here was to have one option somewhere (e.g. in `xportr_write()` where you set the limit as an argument and then it automatically does the split according to some default split rule according to the threshold you provide).

- for the split rules they said doing this by a variable (e.g. `LBCAT`) is impractical as it would end up with a large number of split datasets and you'd have to explain in reviewer's guide exactly how you choose to split each dataset. the convention they usually use instead is to just split by records according to the sort order of the dataset (with as many as possible records being included in split 1 up until the threshold the user provides). then in the reviewers guide you just give one simple sentence to explain how this is done for every dataset where needed.

p.s. something that was news to me... they said that actually they always send via esub both the full oversize datasets and the split datasets bizarrely, and the HA then can just choose either to set them together or use the full datasets provided anyway. makes you wonder why this is even needed in the first place! 

FYI @bundfussr

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

Here's an example of how this could look in future:

`xportr_write(sizelimit=4)`

And within this function have code to check if _sizelimit_ not NULL then apply `xportr_split()`

But default of `xportr_split()` here would be to use the _sizelimit_ user provides and split by records, not by _split_by_

i.e. in this case include the maximum number of records possible in the split 1 dataset up to <=4GB file size

### PR DESCRIPTION
**Pull request by adchan11**
### Thank you for your Pull Request!

We have developed a Pull Request template to aid you and our reviewers. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the xportr codebase remains robust and consistent.

### The scope of `{xportr}`

`{xportr}`'s scope is to enable R users to write out submission compliant `xpt` files that can be delivered to a Health Authority or to downstream validation software programs. We see labels, lengths, types, ordering and formats from a dataset specification object (SDTM and ADaM) as being our primary focus. We also see messaging and warnings to users around applying information from the specification file as a primary focus. Please make sure your Pull Request meets this **scope of {xportr}**. If your Pull Request moves beyond this scope, please get in touch with the `{xportr}` team on [slack](https://pharmaverse.slack.com/archives/C030EB2M4GM) or create an issue to discuss.

Please check off each task box as an acknowledgment that you completed the task. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

### Changes Description

Updated xportr_write with new function and parameter to split data frame based on specified maximum file size. Also updated unit testing to test ability to split data frame and ensure file size of exported files. 

### Task List

- [x] The spirit of xportr is met in your Pull Request
- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Summary of changes filled out in the above Changes Description. Can be removed or left blank if changes are minor/self-explanatory.
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Use `styler` package and functions to style files accordingly.
- [x] New functions or arguments follow established convention found in the [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Roxygen-Headers).
- [x] Updated relevant unit tests or have written new unit tests. See our [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Unit-Tests) for conventions used in this package.
- [x] Creation/updated relevant roxygen headers and examples. See our [Wiki](https://github.com/atorus-research/xportr/wiki/Style-Guide-for-Roxygen-Headers) for conventions used in this package.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Run `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new/updated functions occur on the "Reference" page.
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] The `NEWS.md` entry should go under the `# xportr development version` section. Don't worry about updating the version because it will be auto-updated using the `vbump.yaml` CI.
- [x] Address any updates needed for vignettes and/or templates.
- [x] Link the issue Development Panel so that it closes after successful merging.
- [x] The developer is responsible for fixing merge conflicts not the Reviewer.
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by adchan11** [comment]
Hi @bms63 , I pushed an intial draft of an updated xportr_write function. My unit test in test-write.R is failing the CMD check but when I run it in my local R studio, it passes so I was wondering if you could take a look into this during your review? Thanks.

**Comment by bms63** [comment]
Is there some setting that needs to be set here with the testing large files??  Maybe this shouldn't be enabled for CRAN...not sure if they would allow it??

image

**Comment by adchan11** [comment]
> Is there some setting that needs to be set here with the testing large files?? Maybe this shouldn't be enabled for CRAN...not sure if they would allow it??
> 
> image

Hi @bms63 , I'm not sure about the test_large_files test since I didn't code it and it was already set to not TRUE. In my unit test, I used the adlb dataset from pharmaverseadam and I suspect that is the reason why it's failing the CMD check but I included pharmaverseadam in the DESCRIPTION file and also referenced library(pharmaverseadam) in the testthat.R file.

File: tests/testthat.R
Comment (bms63) [inline tests/testthat.R:2]: I don't believe that this needs to be here

Comment (adchan11) [reply tests/testthat.R]: Thanks @bms63. I've removed this line and pushed.

**Comment by bms63** [comment]
@adchan11 I'm not sure on this failing windows check.

@averissimo @vedhav @elimillera @EeethB do you all see any issues happening in this check?  thanks for taking a peek if you are able!!

File: R/write.R
Comment (vedhav) [inline R/write.R]: Getting the directory path using string manipulation fails in Windows machines because in Windows the directory separators are different. I just pushed a commit to fix this which will eventually fix the CI checks.

**Comment by bms63** [comment]
amazing!! thanks @vedhav !!

**Comment by adchan11** [comment]
> amazing!! thanks @vedhav !!

Thanks @vedhav ! Just wanted to follow up @bms63, if there are any other changes needed, now that all CMD checks passed?

**Comment by bms63** [comment]
I think we are good - @elimillera can you send this to CRAN?  do you want it in main or in this branch?

File: DESCRIPTION
Comment (rossfarrugia) [inline DESCRIPTION:45]: i don't think this is needed

File: NEWS.md
Comment (rossfarrugia) [inline NEWS.md:5]: shouldn't the DESCRIPTION file version update to 0.5.0 and then include this new feature under a new section: 

```
# xportr 0.5.0

## New Features

```
Especially if this is the only update before you'll send back to CRAN.

**Review by rossfarrugia — CHANGES_REQUESTED** [review]

File: DESCRIPTION
Comment (bms63) [reply DESCRIPTION:45]: I think the data from the package is used in the tests now - so it is needed

File: NEWS.md
Comment (bms63) [reply NEWS.md:5]: umm..yeah that should go above 0.4.0!! nice catch  we should probably just change to 0.5.0 and add skip vbump in the PR title.  @elimillera WDYT

File: DESCRIPTION
Comment (rossfarrugia) [reply DESCRIPTION:45]: yep, sorry i missed that!

**Comment by elimillera** [comment]
> I think we are good - @elimillera can you send this to CRAN? do you want it in main or in this branch?

Sounds good. I think we can push to the main branch then I'll push out and make the release once its accepted. I'll also do a couple of checks today to make sure we're all set.

File: NEWS.md
Comment (elimillera) [reply NEWS.md:5]: I agree. This adds new functionality so updating the minor version would be the way I would go

File: R/write.R
Comment (bundfussr) [inline R/write.R:159]: If the size of the xpt file doesn't exceed the maximum size, no counter should be appended to the filename.

The results of the function are correct. However, the performance could be improved. If the file size is smaller than the maximum size, the file is written `log(nr_rows)` times.

Maybe we could write the complete file first and if it exceeds the maximum size, an estimate for the number of rows is calculated based on the file size and the number of rows. If one of the parts exceeds the maximum size, the estimate is adjusted.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: NEWS.md
Comment (adchan11) [reply NEWS.md:5]: Thanks, updated.

File: R/write.R
Comment (adchan11) [reply R/write.R]: Thanks, updated.

File: R/write.R
Comment (adchan11) [reply R/write.R:52]: Thanks, updated.

File: R/write.R
Comment (adchan11) [reply R/write.R:165]: Thanks, updated.

File: R/write.R
Comment (adchan11) [reply R/write.R:110]: If agreed by the team, I can remove it. @bms63 @elimillera

File: R/write.R
Comment (bms63) [inline R/write.R:79]:
```suggestion
assert_numeric(max_size_gb, null.ok = TRUE)
```

File: R/write.R
Comment (bms63) [reply R/write.R:110]: ah dang....that is good catch @bundfussr - we should definitely deprecate it as we release in 0.4 which was 6 months ago.

Can you set this up @adchan11 and recommend folks use `xportr_write()` please

File: R/write.R
Comment (adchan11) [reply R/write.R]: Thanks, updated

File: R/write.R
Comment (adchan11) [reply R/write.R:110]: Hi @bms63 , I updated the NEWS.md with an additional note. Is there anything else I need to do to 'deprecate' the `xportr_split()` function?

File: R/write.R
Comment (bms63) [reply R/write.R]: It needs to get placed in this file with its other friends - https://github.com/atorus-research/xportr/blob/main/R/xportr-package.R

**Comment by bms63** [comment]
@bundfussr are you happy with your requested changes?

File: R/write.R
Comment (bundfussr) [reply R/write.R:165]: I'm not sure if we should optimize the code. For testing I used an ADLB dataset which written as xpt file has 400MB. I wrote the dataset with different values for `max_size_gb`:
```
> system.time(xportr_write(adlb, "adlb.xpt"))
   user  system elapsed 
 25.730   2.258  27.857 
> system.time(xportr_write(adlb, "adlb.xpt", max_size_gb = 1))
Data frame exported to 1 xpt files.
   user  system elapsed 
 50.154  22.826  73.497 
> system.time(xportr_write(adlb, "adlb.xpt", max_size_gb = 0.3))
Data frame exported to 2 xpt files.
   user  system elapsed 
247.439  25.566 275.742 
```
@adchan11 , @bms63 , @rossfarrugia , what do you think?

Comment (rossfarrugia) [reply R/write.R:165]: Less of a concern from my side as most commonly would be ran with 4, 5, or 10 (so creating less splits than your examples above) and usually just a one time job at the end only impacting certain large datasets, so not like people will be running this often.

**Comment by bundfussr** [comment]
> @bundfussr are you happy with your requested changes and thanks for being so thorough!!

I don't know which deprecation strategy you use in xportr. However, in addition to the note I would add a note to the `xportr_split()` documentation and issue an error if `xport_split()` is called as the functionality is no longer available.

File: R/write.R
Comment (adchan11) [reply R/write.R]: Thanks, updated.

**Comment by adchan11** [comment]
> > @bundfussr are you happy with your requested changes and thanks for being so thorough!!
> 
> I don't know which deprecation strategy you use in xportr. However, in addition to the note in the Changelog I would add a note to the `xportr_split()` documentation and issue an error if `xport_split()` is called as the functionality is no longer available.

Thanks, I will work on this over the next few days.

**Comment by adchan11** [comment]
> > @bundfussr are you happy with your requested changes and thanks for being so thorough!!
> 
> I don't know which deprecation strategy you use in xportr. However, in addition to the note in the Changelog I would add a note to the `xportr_split()` documentation and issue an error if `xport_split()` is called as the functionality is no longer available.

Hi @bundfussr, I pushed updates to deprecate the function and improve documentation. I took some guidelines on how to deprecate functions using these sources as I haven't done this before:

https://github.com/tidyverse/dplyr/blob/HEAD/R/deprec-lazyeval.R
https://contributions.bioconductor.org/deprecation.html
https://dplyr.tidyverse.org/reference/se-deprecated.html

Hopefully this follows best practices for deprecation but please let me know your feedback. Thanks.

**Comment by bms63** [comment]
> > > @bundfussr are you happy with your requested changes and thanks for being so thorough!!
> > 
> > 
> > I don't know which deprecation strategy you use in xportr. However, in addition to the note in the Changelog I would add a note to the `xportr_split()` documentation and issue an error if `xport_split()` is called as the functionality is no longer available.
> 
> Hi @bundfussr, I pushed updates to deprecate the function and improve documentation. I took some guidelines on how to deprecate functions using these sources as I haven't done this before:
> 
> https://github.com/tidyverse/dplyr/blob/HEAD/R/deprec-lazyeval.R https://contributions.bioconductor.org/deprecation.html https://dplyr.tidyverse.org/reference/se-deprecated.html
> 
> Hopefully this follows best practices for deprecation but please let me know your feedback. Thanks.

sorry @adchan11 we have a deprecation process in our Wiki https://github.com/atorus-research/xportr/wiki/Deprecation-Process.  It is similar to admiral's process

**Comment by adchan11** [comment]
> > > > @bundfussr are you happy with your requested changes and thanks for being so thorough!!
> > > 
> > > 
> > > I don't know which deprecation strategy you use in xportr. However, in addition to the note in the Changelog I would add a note to the `xportr_split()` documentation and issue an error if `xport_split()` is called as the functionality is no longer available.
> > 
> > 
> > Hi @bundfussr, I pushed updates to deprecate the function and improve documentation. I took some guidelines on how to deprecate functions using these sources as I haven't done this before:
> > https://github.com/tidyverse/dplyr/blob/HEAD/R/deprec-lazyeval.R https://contributions.bioconductor.org/deprecation.html https://dplyr.tidyverse.org/reference/se-deprecated.html
> > Hopefully this follows best practices for deprecation but please let me know your feedback. Thanks.
> 
> sorry @adchan11 we have a deprecation process in our Wiki https://github.com/atorus-research/xportr/wiki/Deprecation-Process. It is similar to admiral's process

Thanks @bms63. I've pushed updates as per the deprecation process you've linked. The CMD checks now pass except one regarding validating links. I reviewed it and the links work for me but I'm not sure why they're not rendering properly in the vignettes. It's only the fda.gov links.

FYI: @bundfussr ready for review if you have any additional feedback. 

Please note, I will be OOO for the next 2 weeks and back on Sept 30. I will review again on Sept 30 when I'm back if there are any changes I need to push. FYI @rossfarrugia

File: R/write.R
Comment (bms63) [reply R/write.R:165]: I'm not super worried about optimizing either.  Hopefully, xpts go away in the next couple of years!

File: R/split.R
Comment (bms63) [inline R/split.R:44]: love this!  ty for doing this!

**Review by bms63 — APPROVED** [review]
This looks good to me - we can finish up when you return @adchan11 unless @rossfarrugia you need this now?

**Comment by rossfarrugia** [comment]
> This looks good to me - we can finish up when you return @adchan11 unless @rossfarrugia you need this now?

Should be fine for us, thanks! FYI @millerg23 - If this goes back to CRAN early Oct then we should be able to squeeze in implementing it with our next admiralroche release. Although if @bundfussr agrees his comments are resolved now then we wouldn't need to wait for Adrian and could push ahead anytime.

**Review by bms63 — COMMENTED** [review]
@rossfarrugia  if needed, I can try and do final review EOW and we can push to CRAN.  @elimillera do you have time to review as well?  

Honestly, feels like this is in pretty good shape

**Comment by rossfarrugia** [comment]
thanks @bms63 - if no extra review comments needing Adrian's input then yes i'd agree. that'd help us as we do have one team already lined up that will be wanting to use this new functionality.

File: R/split.R
Comment (bundfussr) [reply R/split.R:44]: Usually, the deprecated functionality is still available in Phase 1 (deprecation warning). However, here `xportr_split()` still works as before but in `xportr_write()` it is ignored. I.e., the functionality is already removed.

I would either go directly to Phase 2 (deprecation error) or keep the old functionality (in addition to the new functionality) in `xport_write()`. I assume `xportr_split()` is used only by a few users (if at all). Thus I would tend to the first option.

@bms63 , what do you think?

**Comment by bundfussr** [comment]
> Although if @bundfussr agrees his comments are resolved now then we wouldn't need to wait for Adrian and could push ahead anytime.

The new functionality is OK from my side. Just the deprecation of the old functionality deviates from the usual process.

**Comment by bms63** [comment]
whew!!! two weeks goes fast.   this is a perpetual item on my todo list.  Still trying to get to it!

**Comment by rossfarrugia** [comment]
@bms63 - @bundfussr has offered to make a commit to make the final updates for his review comments. Hopefully this'll help in pushing the release out and getting it off to CRAN

**Comment by bundfussr** [comment]
@bms63 , I have a commit ready for finalizing the deprecation. Could you grant we write access to the repo? Then I can push it.

**Review by bundfussr — APPROVED** [review]

**Comment by bundfussr** [comment]
@bms63 , should we create a `.lycheeignore` file to fix the failing links check or should we just ignore it?

**Comment by bms63** [comment]
Bummer looks like those Phuse sites are dead now...they were very helpful.

**Pull request merged**
